core/src/main/scala/scala/pickling/generator/PicklingMacros.scala

```diff
@@ -0,0 +1,129 @@
+package scala.pickling
+package generator
+
+import scala.collection.mutable
+
+import HasCompat._
+
+private[pickling] trait PicklingMacros
+    extends Macro with SourceGenerator with TypeAnalysis {
+
+  implicit val implContext = c
+
+  import c.universe._
+  import compat._
+
+  val symbols = new IrScalaSymbols[c.universe.type, c.type](c.universe, tools)
+  // TODO - also allow import to disable the warnings.
+  val disableWillRobinsonWarning =
+    java.lang.Boolean.getBoolean("pickling.willrobinson.disablewarning")
+  // TODO - We should have more customization than this
+  val handleCaseClassSubclasses = !configOption(
+      typeOf[IsIgnoreCaseClassSubclasses])
+  val generator =
+    if (isStaticOnly) {
+      PicklingAlgorithm.aggregate(
+          Seq(new CaseClassPickling(
+                  allowReflection = false,
+                  careAboutSubclasses = handleCaseClassSubclasses),
+              AdtPickling,
+              ScalaSingleton))
+    } else {
+      PicklingAlgorithm.aggregate(
+          Seq(new CaseClassPickling(
+                  allowReflection = true,
+                  careAboutSubclasses = handleCaseClassSubclasses),
+              AdtPickling,
+              ScalaSingleton,
+              new ExternalizablePickling,
+              new WillRobinsonPickling(
+                  showWarnings = !disableWillRobinsonWarning)))
+    }
+
+  implicit object logger extends AlgorithmLogger {
+    def warn(msg: String): Unit = c.warning(c.enclosingPosition, msg)
+    def error(msg: String): Unit = c.error(c.enclosingPosition, msg)
+    def abort(msg: String): Nothing = c.abort(c.enclosingPosition, msg)
+    def debug(msg: String): Unit = {
+      // These are enabled when -verbose is enabled.
+      c.info(c.enclosingPosition, msg, force = false)
+    }
+  }
+
+  import MacrosErrors._
+
+  def checkClassType(tpe: c.Type, isUnpickle: Boolean = true): Unit = {
+    import definitions._
+    tpe.normalize match {
+      case tpe1 if tpe1.typeSymbol.isClass => () // This case is fine.
+      case NothingTpe => impossibleGeneration("Nothing")
+      case RefinedType(parents, decls) => impossibleGeneration("refined type")
+      case _ if tpe.isScalaOrJavaPrimitive =>
+        impossibleGeneration("primitive types")
+      case _ => impossibleGeneration(s"non-class type $tpe")
+    }
+  }
+
+  def preferExistingImplicits(body: => Tree): Tree = {
+
+    import Compat._
+    import Console._
+
+    val candidates = c.enclosingImplicits
+    if (candidates.isEmpty) return body
+    val ourPt = candidates.head.pt
+
+    val key = ourPt.toString
+    Memo.memo += key -> (Memo.memo.getOrElse(key, 0) + 1)
```

Comment:
Let's leave it for a moment, I want to play around with the implicit search a little bit more. It's only used at compile-time, in the final release will be removed :+1:  Is this ok to be merged?